common/channelconfig/util.go

```diff
@@ -241,51 +242,72 @@ func ACLValues(acls map[string]string) *StandardConfigValue {
 
 // ValidateCapabilities validates whether the peer can meet the capabilities requirement in the given config block
 func ValidateCapabilities(block *cb.Block, bccsp bccsp.BCCSP) error {
+	cc, err := extractChannelConfig(block, bccsp)
+	if err != nil {
+		return err
+	}
+	// Check the channel top-level capabilities
+	if err := cc.Capabilities().Supported(); err != nil {
+		return err
+	}
+
+	// Check the application capabilities
+	return cc.ApplicationConfig().Capabilities().Supported()
+}
+
+// ExtractMSPIDsForApplicationOrgs extracts MSPIDs for application organizations
+func ExtractMSPIDsForApplicationOrgs(block *cb.Block, bccsp bccsp.BCCSP) ([]string, error) {
+	cc, err := extractChannelConfig(block, factory.GetDefault())
+	if err != nil {
+		return nil, err
+	}
+
+	if cc.ApplicationConfig() == nil {
+		return nil, errors.Errorf("could not get application config for channel")
+	}
+	orgs := cc.ApplicationConfig().Organizations()
+	mspids := make([]string, 0, len(orgs))
+	for _, org := range orgs {
+		mspids = append(mspids, org.MSPID())
+	}
+	return mspids, nil
+}
+
+func extractChannelConfig(block *cb.Block, bccsp bccsp.BCCSP) (*ChannelConfig, error) {
 	envelopeConfig, err := protoutil.ExtractEnvelope(block, 0)
 	if err != nil {
-		return errors.Errorf("failed to %s", err)
+		return nil, errors.Errorf("failed to %s", err)
 	}
 
 	configEnv := &cb.ConfigEnvelope{}
 	_, err = protoutil.UnmarshalEnvelopeOfType(envelopeConfig, cb.HeaderType_CONFIG, configEnv)
 	if err != nil {
-		return errors.Errorf("malformed configuration envelope: %s", err)
+		return nil, errors.Errorf("malformed configuration envelope: %s", err)
 	}
 
 	if configEnv.Config == nil {
-		return errors.New("nil config envelope Config")
+		return nil, errors.New("nil config envelope Config")
 	}
 
 	if configEnv.Config.ChannelGroup == nil {
-		return errors.New("no channel configuration was found in the config block")
+		return nil, errors.New("no channel configuration was found in the config block")
 	}
 
 	if configEnv.Config.ChannelGroup.Groups == nil {
-		return errors.New("no channel configuration groups are available")
+		return nil, errors.New("no channel configuration groups are available")
 	}
 
 	_, exists := configEnv.Config.ChannelGroup.Groups[ApplicationGroupKey]
 	if !exists {
-		return errors.Errorf("invalid configuration block, missing %s "+
+		return nil, errors.Errorf("invalid configuration block, missing %s "+
 			"configuration group", ApplicationGroupKey)
 	}
 
 	cc, err := NewChannelConfig(configEnv.Config.ChannelGroup, bccsp)
 	if err != nil {
-		return errors.Errorf("no valid channel configuration found due to %s", err)
-	}
-
-	// Check the channel top-level capabilities
-	if err := cc.Capabilities().Supported(); err != nil {
-		return err
+		return nil, errors.Errorf("no valid channel configuration found due to %s", err)
```

Comment:
Nit: this isn't a very idiomatic error statement, in general we would wrap the context around the error, not use phrasing like "due to"

Reply:
Moreover, the use of `errors.Errorf` will cause loosing the stack trace of the actual underlying error. On a general pattern for using `"github.com/pkg/errors"`, we should use
- `New / Errorf`, when constructing the original error in the fabric code (so that the stack trace is generated)
- `Wrapf` when propagating the error received from an external library (so that stack trace is generated)
- Finally, for propagating the errors from fabric apis (as is the case here, and majority of the fabric code), we should either forward as is, if we don't have much to add to the context. If we want to add some context, we should use `WithMessagef` so that the full stack trace is not lost.

Reply:
@btl5037 @manish-sethi Thank you. This method was simply split from the existing ValidateCapabilities method for code sharing. Therefore, all the error handlings were kept unchanged from the original code. Having said so, definitely agree with your comments. Will fix it.